### ISSUE TITLE
Questions around speedup

### ISSUE THREAD

**Issue by 1m1ne** [deepmind/jmp#28]
Hi,

Thanks for creating this amazing library!

So from what I understood this is the minimal needed to benefit from the JMP speedup
```python
policy = jmp.Policy(param_dtype=jnp.float32, compute_dtype=jnp.float16, output_dtype=jnp.float16)
...
# creating a network, and creating a loss_fn using the network
data = policy.cast_to_compute(data)
params = policy.cast_to_compute(params)

grads = jax.grad(loss_fn)(data, params)
...
grads = policy.cast_to_param(grads)
```
The `loss scale` is only needed if we experience `NaN` or `inf` values.
And thus we should see a difference in the time needed for a `jax.grad` operation when the data and params are in `float16`.

Please correct me if I'm wrong or if I miss anything.

Considering that, I have some questions:
* Can we expect to see a 2x or any speedup if we time the `jax.grad` operation with (params, data) in `float16` and in `float32` or is the speedup only achieved at scale ?
* Can we expect to see a 2x or any speedup with small networks and small experiments ? (say 2-layer nets and experiments that only need 2-3 minutes on a single gpu)
* Can we expect to see a 2x or any speedup with all types of networks or only some with specific architectures ? (say a 50 layer MLP, Transformers, etc)
* Is it necessary to apply the mixed precision policy to the network and thus to use `Haiku` for `hk.mixed_precision.set_policy` or is having the parameters and data in `float16` sufficient to have a speedup even with a network created by us ?

Thank you and have a nice day !

**Comment by 1m1ne** [comment]
@tomhennigan :smiley: :smiley: :smiley: Would be very happy to know what you think ?

**Comment by tomhennigan** [comment]
> So from what I understood this is the minimal needed to benefit from the JMP speedup

This would let you observe the speedup from doing training for everything in your step in half precision.

It is also sometimes necessary to upcast certain parts of the network (e.g. BatchNorm) to full precision for numerical stability.

> The loss scale is only needed if we experience NaN or inf values.

I would expect if you are using `float16` (not `bfloat16`) then you will need loss scaling to be able to train anything interesting.

If you have access to a TPU, or to a GPU with `bfloat16` support (e.g. an NVIDIA A100) then you should be able to use `bfloat16` as the compute dtype and avoid loss scaling.

> Can we expect to see a 2x or any speedup if we time the jax.grad operation with (params, data) in float16 and in float32

It depends on your model code. Most accelerators have significantly better (like 2-10x) performance in half precision, so you may see >2x.

In https://github.com/deepmind/dm-haiku/commit/dba1fd9c3776393b35b49cd319b31375975f5a97 I added mixed precision support to our Haiku imagenet training script and observed a 2x speedup on V100 GPUs.

> is the speedup only achieved at scale ?

You should see speedups at single GPU scale (from the improvement to compute performance) and if you keep your grads in half precision you will see an improvement at multi-host scale (due to a reduction in the amount of data being sent).

> Can we expect to see a 2x or any speedup with small networks and small experiments ? (say 2-layer nets and experiments that only need 2-3 minutes on a single gpu)

Maybe. For tiny networks often there are often other overheads that dominate (e.g. the time taken to transfer batches of data from host->device, or python overheads in your training script).

> Can we expect to see a 2x or any speedup with all types of networks or only some with specific architectures ? (say a 50 layer MLP, Transformers, etc)

The improved performance comes from accelerators having specialised hardware for doing matrix multiplication (and conv) in half precision. So any workloads that are dominated by matrix multiply would be faster.

> Is it necessary to apply the mixed precision policy to the network and thus to use Haiku for hk.mixed_precision.set_policy or is having the parameters and data in float16 sufficient to have a speedup even with a network created by us ?

The functions in Haiku are designed to help selectively apply policies to different modules. For example in resnet you probably want to keep BatchNorm in full precision but use half precision elsewhere. They are just a convenience mechanism for calling `policy.cast_to_{x}` on your behalf, they are not required for using `jmp`.

**Comment by 1m1ne** [comment]
@tomhennigan Thanks a lot for all your detailed answers, it's a lot more clear now !
